File: comments.py
# -*- coding: UTF-8 -*-
import sys
from time import sleep
if sys.version_info[0]==2: #Вот это нужно, чтобы под вторым питоном работал UTF-8
	reload(sys)
	sys.setdefaultencoding('utf-8')
import tabun_api as api
import re
from random import randint
from os.path import exists
user = api.User(login='', passwd='') #Вход
def check(comms): #Получает аргументом список комментов и делает главные штуки
	for c in comms.items():
		comm_id=c[0]
		c=c[1]
		post_id=c.post_id
		was_answered=False #Чтобы не повторяться тут есть эта переменная
		if re.search("@er16 ignore", c.raw_body):
			"""
			Вообще, эта штука переключает игнор
			Ну и тут захардкожено, что файл находится в ./tabun/
			Так что нужно запускать прогу на уровень выше,
			аля python tabun/comments.py
			Тогда игнор будет работать =3
			"""
			for isansw in comms.items():
				answ_id=isansw[0]
				isansw=isansw[1]
				if isansw.parent_id==comm_id and isansw.author==user.username:
					was_answered=True #Вот тут проверка, ответили ли
			if not was_answered:
				"""
				Тут много костылей с файлами
				Ибо оно как-то странно записывает
				Но эт скорее всего я криворукий
				Я ещё за эту часть мне стыдно, да.
				"""
				f=open("tabun/ignore.lst","r+")
				if c.author.strip() in f.read().split("\n"):
					f.close();f=open("tabun/ignore.lst","r") #Если уже есть в списке - убрать
					cutted=re.sub(c.author.strip(),"",f.read())
					f.close()
					f=open("tabun/ignore.lst","w")
					f.write(cutted)
					user.comment(target_id=c.post_id,body="Игнорирование выключено",
							typ="blog",
							reply=comm_id)
				else:
					f.close()
					f=open("tabun/ignore.lst","a") #Ну и соответственно наоборот
					f.write(c.author+"\n")
					user.comment(target_id=c.post_id,body="Игнорирование включено",
							typ="blog",
							reply=comm_id)
				f.close()
				continue

		if re.search("<pre>[\s\S]*</pre>",c.raw_body):
			c.raw_body=re.sub("<pre>[\s\S]*</pre>","",c.raw_body)
		if re.search("<blockquote>[\s\S]*</blockquote>",c.raw_body):
			c.raw_body=re.sub("<blockquote>[\s\S]*</blockquote>","",c.raw_body)
		"""
		^ Вот эти двое вырезают тэги pre и blockquote,
		Там такое неуместно, ибо с о б а ч к и
		"""

		if re.search("@(?P<name>\w+)",c.raw_body): #Главная часть с упоминаниями
			username=list(set(re.findall('@(?P<name>\w+)',c.raw_body)[:10])) #Обрезать до 10 и убрать повторы
			if "am31" in username or "lunabot" in username or "autopilot" in username or "er16" in username: #Не трогать комменты с ботами
				continue
			existing_users=[]
			f=open("tabun/ignore.lst")
			ignored_names=f.read().split("\n") #Получить игнорлист
			f.close()
			for ch in username:
				"""
				Проверяет существование юзера посредством пингования страниц по очереди
				Если выплёвывает ошибку - пропускает
				"""
				try:
					name=user.get_profile(username=ch).username
					if name not in ignored_names:
						existing_users.append(name)
				except api.TabunError:
					pass
			f.close()
			username=existing_users; del existing_users;  # За эту часть мне тоже стыдно, просто так было проще :3
			for isansw in comms.items(): #Тут проверка на ответы
				answ_id=isansw[0]
				isansw=isansw[1]
				if isansw.parent_id==comm_id and (isansw.author==user.username or isansw.author==username):
					was_answered=True
			if not was_answered:
				"""
				Тут генерирует само сообщение
				Я жестко юзал format, ибо нормального
				вставления переменных в строки
				(Слава руби!) нету
				"""
				body=("""Вас упомянул в посте <a href='{post_url}'>'{post_name}'</a> \
		пользователь <ls user="{user_name}" />
		<a href='{comm_url}'>Ссылка</a> на комментарий""".format(
					post_url='/blog/{}/{}.html'.format(c.blog,c.post_id),
					post_name=user.get_post(post_id).title,
					user_name=c.author,
					comm_url="/blog/{blog}/{post_id}.html#comment{comm_id}".format(
						blog=c.blog,
						post_id=c.post_id,
						comm_id=c.comment_id)))
				if len(username)>=1:  #Если есть кому отправить - сделать это
					user.add_talk(talk_users=username,
					title="Упоминание",
					body=body)
				if len(username)>1: #Ну тут уже чисто чтобы выглядело нормально, если больше одного, то пересислить после двоеточия
					cbody="Сообщение об упоминании отправлено пользователям: "
					for x in username:
						cbody+="<a href='/profile/{}'>".format(x)+x+"</a>, "
					cbody=cbody[:-2]
				elif len(username)==1: #Если нет, то просто сказать
					cbody="Сообщение об упоминании отправлено пользователю <a href='/profile/{0}'>{0}</a>".format(username[0])
				else:
					cbody="Ни одного из пользователей, упомянутых в сообщении, не существует или все из них включили игнорирование."
				user.comment(target_id=c.post_id,
				body=cbody,
				typ="blog",
				reply=comm_id) #Доложить, что всё сделано
def run():
	while True: #Бесконечный цикл, чтобы не запускать кроном
		try:
			check(user.get_comments()) #Берёт комменты
			sleep(15) #По идее, чтобы табун не пинал, но я не знаю, пинает или нет :3
		except api.TabunError:
			run()
run()

File: comments.rs
extern crate regex;

use libtabun::{TClient,Comment,CommentType,HOST_URL};

use std::collections::HashMap;
use std::{fs,thread,time};
use std::io::{Read,Write};

use regex::Regex;

fn check(user: &mut TClient, comms: HashMap<i64,Comment>) {
    let pre_regex = Regex::new(r"<pre>[\s\S]*</pre>").unwrap();
    let quote_regex = Regex::new(r"<blockquote>[\s\S]*</blockquote>").unwrap();
    let name_regex = Regex::new(r"@(?P<name>[\w\\-]+)").unwrap();

    for comm in comms.values().by_ref() {
        let mut body = comm.body.clone();
        let was_answered = comms.iter().by_ref().any(|(_,ref x)| x.parent == comm.id && x.author == user.name);
        if body.contains("@er16 ignore") {
            if !was_answered {
                let mut content = String::new();
                {
                    let mut f = fs::File::open("ignore.lst").unwrap();
                    let _ = f.read_to_string(&mut content);
                }
                let ignore_lst = content.split("\n").collect::<Vec<_>>();
                match ignore_lst.iter().position(|&x| x == comm.author) {
                    Some(x) => {
                        let mut new_content = ignore_lst.clone();
                        let _ = new_content.remove(x);
                        let mut f = fs::File::create("ignore.lst").unwrap();
                        let _ = new_content.iter().map(|&x| writeln!(&mut f,"{}", x));
                        let _ = f.sync_data();
                        let _ = user.comment(comm.post_id, "Игнорирование выключено", comm.id, CommentType::Post);
                    }
                    None => {
                        let mut f = fs::OpenOptions::new()
                            .append(true).open("ignore.lst").unwrap();
                        let _ = f.write_all(format!("\n{}", comm.author).as_bytes());
                        let _ = f.sync_data();
                        let _ = user.comment(comm.post_id, "Игнорирование включено", comm.id, CommentType::Post);
                    }
                }
                continue
            }
        }

        if pre_regex.is_match(&body) {
            body = pre_regex.replace_all(&body, "");
        }
        if quote_regex.is_match(&body) {
            body = pre_regex.replace_all(&body, "");
        }
        if name_regex.is_match(&body) {
            if !was_answered {
                let mut names = name_regex.captures_iter(&body).map(|x| x.name("name").unwrap()).take(10).collect::<Vec<_>>();
                names.sort();
                names.dedup();
                if names.iter().any(|x| **x == "lunabot".to_owned() 
                                    || **x == "autopilot".to_owned() 
                                    || **x == "er16".to_owned() 
                                    || **x == "am31".to_owned()) { continue }

                let mut f = fs::OpenOptions::new()
                    .read(true)
                    .write(true)
                    .create(true).open("ignore.lst").unwrap();
                let mut content = String::new();
                let _ = f.read_to_string(&mut content);
                let ignored_names = content.split("\n").collect::<Vec<_>>();

                let names = names.iter().fold(Vec::new(),|mut acc,&x| {
                    match user.get_profile(x) {
                        Ok(x)   => if !ignored_names.iter().any(|y| *y == x.username) { acc.push(x.username) },
                        _       => ()
                    }
                    acc
                });

                let body = format!("Вас упомянул в посте <a href='{post_url}'>'{post_name}'\
                                       </a> пользователь <ls user='{user}' /> \n <a href='{comm_url}'>Ссылка</a> на комментарий",
                                       post_url     = format!("{}/blog/{}.html", HOST_URL, comm.post_id),
                                       post_name    = user.get_post("",comm.post_id).unwrap().title,
                                       user         = comm.author,
                                       comm_url     = format!("{}/blog/{}.html#comment{}",HOST_URL, comm.post_id, comm.id));
                if names.len() >= 1 {
                    let tid = user.add_talk(&names.iter().map(|x| x.as_str()).collect(), "Упоминание", &body).unwrap();
                    let _ = user.delete_talk(tid);
                }
                let cbody = if names.len() > 1 {
                    let mut c = format!("Сообщение об упоминании отправлено пользователям: {}",
                                        names.iter().fold(String::new(),|mut acc, x| { acc.push_str(&format!("<ls user='{}'>, ",x)); acc }));
                    c.pop(); c.pop();
                    c
                } else if names.len() == 1 {
                    format!("Сообщение об упоминании отправлено пользователю <ls user='{}'>",names[0])
                } else {
                    String::new()
                };

                if !cbody.is_empty() {
                    let _ = user.comment(comm.post_id, &cbody, comm.id, CommentType::Post);
                }
            }
        }
    }
}

pub fn dothing() {
    let mut user = TClient::new("", "").unwrap();
    loop {
        let comms = user.get_comments("").unwrap();
        check(&mut user, comms);
        thread::sleep(time::Duration::new(30,0))
    }
}
